app/models/work_packages/scopes/relatable.rb

```diff
@@ -0,0 +1,207 @@
+      #   or transitive ancestor/descendant relationship. So a relation between siblings is just as possible as one between
+      #   aunt and nephew.
+      # * No circles between trees: The ancestor/descendant chain is considered bidirectional when calculating relatability.
+      #   This means that starting from the work package both the descendants as well as the ancestors are considered. That
+      #   way, relations like
+      #     Child 1 --- parent ---> Parent 1 ---> follows ---> Parent 2 ---> child ---> Child 2 --- follows ---> Child 1
+      #   are prevented.
+      #
+      # For the sake of this scope, the parent relation (Relation::TYPE_PARENT) is also included in the list of relation_types
+      # even though it is not stored in the same data structure. All Relation::TYPE_* values can be provided even those that
+      # are not canonical, e.g. Relation::TYPE_PRECEDES. The calculation for those relation types are then inverted and the
+      # canonical type is used, e.g. Relation::TYPE_FOLLOWS.
+      #
+      # There are a couple of exceptions and additions to the limitations outlined above for the following types:
+      # * Relation::TYPE_RELATES: Since this is essentially unrelated and does not carry a lot of semantic, the work packages
+      #   are simply somehow related, such relations do not follow the "non circular" nor the "ancestor/descendant" rule.
+      # * Relation::TYPE_PARENT: Since creating a new relationship will remove the old parent relationship, current ancestors
+      #   (except the direct parent) are relatable to. Descendants however are not since that would create a circle.
+      #
+      # The implementation right away excludes all work packages with which a direct relation already exists
+      # and uses a CTE to find all the work packages with which a transitive relationship based on the rules outlined above
+      # exist.
+      def relatable(work_package, relation_type)
+        return all if work_package.new_record?
+
+        scope = not_having_directed_relation(work_package, relation_type)
+                  .not_having_direct_relation(work_package)
+                  .where.not(id: work_package)
+
+        if Setting.cross_project_work_package_relations
+          scope
+        else
+          scope.where(project: work_package.project)
+        end
+      end
+
+      def not_having_direct_relation(work_package)
+        where.not(id: Relation.where(from: work_package).select(:to_id))
+             .where.not(id: Relation.where(to: work_package).select(:from_id))
+      end
+
+      def not_having_directed_relation(work_package, relation_type)
+        sql = <<~SQL.squish
+          WITH
+            RECURSIVE
+            #{non_relatable_paths_sql(work_package, relation_type)}
+
+            SELECT id
+            FROM #{related_cte_name}
+        SQL
+
+        scope = where("work_packages.id NOT IN (#{sql})")
```

Comment:
Should we add `Arel.sql` here to solve the warning?
```suggestion
        scope = where("work_packages.id NOT IN (#{Arel.sql(sql)})")
```

Reply:
I don't get a warning and think I have them activated. Can you point me to what to do to receive it, please?

Reply:
@ulferts it's here: https://github.com/opf/openproject/security/code-scanning/98

Reply:
@dombesz thanks for the pointer. Fixed in b51a5dde06dc5288509af50a607774815a1485f8